### ISSUE TITLE
Add `inconsistent_default_argument` rule

### ISSUE THREAD

**Issue by rgoldberg** [realm/SwiftLint#6668]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New Rule Request

Please add a `inconsistent_default_argument` rule (it could be named `coordinate_default_arguments` if you want an imperative name for the fix instead of a descriptive name for the problem, or something else).

In certain circumstances, it would ensure that parameters with the same label & type that have a default argument use the same default argument.

It could have a few options to specify the applicable circumstances. I'll leave it to you to decide what circumstances (& if the whole rule) should be enabled by default (also, you can rename any options, add options, etc.) (I'd prefer the rule & all options to be enabled by default, but I don't know how much code would violate the maximal rule, so I don't want to cause a tsunami of violations). Boolean options:

- `call_default_arguments`: If functions F & G both have a default argument for parameter P, if F calls G passing its value of P to to G's P, then it is an error if the default arguments for P for both F & G are not the same.
- `overload_default_arguments`: If methods F & G for the same type with the same name but with different signatures both have a default argument for parameter P, then it is an error if the default arguments for P for both F & G are not the same.
- `override_default_arguments`: If method F exists for type T, and overrides a method G from some type/protocol, S, that T extends / conforms to, and both F & G  have a default argument for parameter P, then it is an error if the default arguments for P for both F & G are not the same.
- etc.

**Comment by nenadvulic** [comment]
I'd like to agree on scope with you @SimplyDanny before writing code, because the three variants differ a lot in feasibility for a standard
  (single-file, syntactic) rule.

  **What's cleanly feasible as a regular rule**

  `overload_default_arguments`: within a single type declaration, find overloaded methods (same base name) and flag when a shared `(label, type)` parameter carries *different*
  default-argument expressions across the overloads. This is purely syntactic, single-file, and bounded: it walks the members of each type, groups by function base name, and
  compares the default-value expressions of matching parameters. Easy to test and to scope.

  **What I think doesn't fit a standard rule**

  - `call_default_arguments` needs to resolve that F actually calls G and passes that parameter through a call graph that's frequently cross-file.
  - `override_default_arguments` needs type-hierarchy resolution (does `T` conform to / extend `S`?), which is semantic information not available to a syntactic rule. This would
  likely require an *analyzer rule* (compiler-backed), a much larger effort.

  **Proposal**

  Ship a v1 as an opt-in rule (off by default to avoid a violation tsunami) covering only the `overload_default_arguments` case  same type, same method name, divergent defaults
  for a shared parameter. The cross-file/semantic variants can be considered later, possibly as an analyzer rule, if there's appetite.

@SimplyDanny does scoping it down this way sound acceptable to you? 
If so, I'll send a focused PR with the rule + tests following the existing rule template.

**Comment by SimplyDanny** [comment]
@nenadvulic: Your breakdown looks reasonable. Especially, `override_default_arguments` cannot be implemented on syntax information only. `call_default_arguments` is achievable with potentially high false positive and negative rates. It can also only reliably be implemented with type information and symbol resolution.

What's left is `overload_default_arguments`. That part seems easy enough to achieve. However, it can also not be complete considering function declarations in extensions. It would trigger for example in:

```swift
struct S {
  func f(name: String = "foo") {}
  func f(a: Int, name: String = "bar") {}
}
```

Can someone provide a good example that shows this is of practical use?

**Comment by nenadvulic** [comment]
@SimplyDanny The value is catching *accidental drift* of a default across overloads that callers reasonably expect to behave the same.

A realistic case:

```swift
extension URLSession {
  func fetch(_ url: URL, retries: Int = 3) { ... }
  func fetch(_ url: URL, headers: [String: String], retries: Int = 0) { ... }
}
```

Someone who has `fetch(url)` (3 retries) and later adds headers silently drops to 0 retries — nothing at the call site signals that. Same shape shows up a lot with things like `duration`, `animated`, `timeout`, `isEnabled`:

```swift
func animate(_ changes: () -> Void, duration: TimeInterval = 0.3) { ... }
func animate(_ changes: () -> Void, duration: TimeInterval = 0.25, completion: () -> Void) { ... }
```

The `0.3` vs `0.25` is almost certainly a typo/drift, and it's exactly the kind of thing review misses because the two overloads are rarely read side by side.

On your example with `name: "foo"` vs `name: "bar"`: I agree that's sometimes intentional, which is why this should be **opt-in (disabled by default)** — teams that want the "overloads share defaults" invariant get it, others aren't
bothered.

On extensions: agreed it can't be complete. I'd treat it as best-effort within the syntactic scope SwiftLint sees (a type body or a single extension), same as other syntactic rules that don't chase declarations across files. False
negatives across extensions are acceptable for an opt-in rule; it still catches the common in-type case shown above.

**Comment by rgoldberg** [comment]
I assume that an analyzer rule could handle all of the scenarios.

If someone implements a limited normal rule, I think that the limitation should be noted in the rule name, both so that users don't expect the rule to do more than it actually does, and so that a complete analyzer rule with the general name can be eventually be made without a name clash.

If default arguments can be checked across all declarations in a file (i.e. across any original type declaration plus any extensions in the same file), then maybe `inconsistent_default_argument_in_file`.

If default arguments can be checked only within a single declaration (i.e. only within the original type declaration or within a single extension), maybe `inconsistent_default_argument_in_declaration`.

**Comment by SimplyDanny** [comment]
Okay. Sounds reasonable, but should definitely be opt-in with the prospect of an Analyzer rule extending this rule restricted in its scope.

**Comment by nenadvulic** [comment]
@SimplyDanny  I'll implement it as an opt-in rule named `inconsistent_default_argument_in_file`, checking across the original type declaration plus any extensions of it in the same file (purely syntactic, single-file). A future Analyzer rule can cover the cross-file cases without a name clash.

it works for you?

**Comment by rgoldberg** [comment]
Looking at it again, `inconsistent_default_argument_in_file` could imply that it's for any parameter with the same name within the same file, regardless if it's for the same type or different types.

I think the name would be fine, and functionality improved, if options like the following are added:

`container_scope` option:

- `across_all`: rule is performed a across everything in the file
- `across_one_type`: rule is performed across a type & all extensions for that type
- `within_one`: rule is performed separately for container scope

Container scopes include global, type declarations, extensions, code blocks (including funcs, computed vars, inits, deinits, code blocks like if, else, do, etc.).

`nested_containers` option:

- `include`: process nested containers along with parent container, like `ParentStruct.NestedStruct` along with `ParentStruct`
- `exclude`: process nested containers separately from parent container

I used `container(s)` in the option names to allow flexibility for potential future options (maybe there'd be an option for a different type of scope), but the word can be omitted if advisable.

If more than one of `call_default_arguments`, `overload_default_arguments` & `override_default_arguments` are supported, they could be either individual Boolean options or values for one array option (maybe named `locations` with values `call`, `overload` & `override`, or `call_site`, `overloaded_function` & `overridden_function`, respectively; the values could be plural).

The options and/or their values could be renamed, augmented, etc.

Thanks for discussing & implementing this.

**Comment by nenadvulic** [comment]
Thanks @rgoldberg, these are good points — especially the ambiguity in the name.

To keep momentum on what @SimplyDanny already approved (an opt-in, single-file syntactic rule, with an Analyzer rule covering the rest later), I'd like to ship a focused v1 first and design it so your options can be layered on without breaking changes.

The scope:
- Name: inconsistent_default_argument_in_file
- Behavior = your container_scope: across_one_type (a type's primary declaration + all its extensions in the same file), grouping by function base name and flagging divergent defaults for a shared (label, type) parameter.
- Opt-in (disabled by default).
- Nested types handled separately for now (nested_containers: exclude), to keep v1 bounded.

Designed for your options as follow-ups:
- container_scope (across_all / across_one_type / within_one) — across_one_type becomes the default, so adding the others is purely additive.
- nested_containers (include / exclude), exclude as default.
- locations (overload for now; call / override reserved for the future Analyzer rule).

This way the rule name and config surface stay forward-compatible with everything you described, and we avoid a name clash with the eventual full Analyzer rule.

@SimplyDanny  does shipping v1 = across_one_type (with the config knobs added incrementally) work for you?

**Comment by rgoldberg** [comment]
Thanks for going thorough all this.

A few more potential future options (not for v1; they might be overkill, but I want to document my ideas if anyone wants them added):

`parameter_grouping`:

- `type`: all parameters with the same type, regardless of name, are grouped together
- `type_and_name` (default): all parameters with the same type & name are grouped together

`function_grouping` (this includes funcs, inits, subscripts, or anything else that can have default arguments, so it should be renamed if someone knows a concise collective term for these):

- `function_kind_groups`: array of arrays containing `init`, `func`, `subscript`, etc. (with each occurring at most once), each child array being a list of "function" kinds whose parameters are processed together. If you can have named values, `all_together` (this would be default) would have exactly one child array with all the kinds in it, while `each_separately` would have one child array for each "function" kind by itself.

`default_required`: Boolean:

- `true`: if one parameter in a group has a default argument, all others in the same group must have the same one (i.e. all or none have a default argument)
- `false` (default): all parameters in the same group that have a default argument must have the same one (i.e. each parameter in a group individually either has no default argument or one that is the same for all parameters in the group)

**Comment by SimplyDanny** [comment]
> [@SimplyDanny](https://github.com/SimplyDanny) does shipping v1 = across_one_type (with the config knobs added incrementally) work for you?

Yes, sounds good all in all. Incremental development is fine as long as we always start with reasonable defaults that can be modified in a later update optionally without breaking what people have already gotten used to (with exceptions of course). The plan you propose seems to follow that goal.

The only thing I wonder about is:

> * Nested types handled separately for now (nested_containers: exclude), to keep v1 bounded.

I understand that this keeps the scope a little smaller. But is this a reasonable default; something users would expect? Nested types are not necessarily inaccessible. In fact, they are often only nested for structural reasons but are otherwise used like any other type. Both `include`/`exclude` are fine for me. I'd like to understand the reason behind that though.

**Comment by nenadvulic** [comment]
Thanks @SimplyDanny! Good question, and I think the wording exclude is what's misleading here.

Under across_one_type, exclude does not mean nested types are skipped. It means each nested type is checked as its own type scope (its primary declaration + its extensions in the file), separately from the parent. So Parent.Nested is still fully covered — just not merged into Parent.

That separation is actually the only correct behavior at this scope: Parent.f(…) and Parent.Nested.f(…) are functions on different types, so grouping them (include) would compare unrelated defaults and produce false positives. 
include only becomes meaningful under across_all, where merging everything in the file is the explicit intent.

So for v1 I'd keep nested types checked-but-isolated, and surface include/exclude as a real choice once across_all lands. Happy to rename if exclude reads as "ignored", maybe documenting it as "nested types form their own scope" is clearer.

**Comment by rgoldberg** [comment]
Good call about `nested_containers` `include` & `exclude`; let's rename them to `with_ancestors` & `separately` (or something similar), respectively.

Thanks again for going over all this theoretical complexity, and for implementing this.

**Comment by SimplyDanny** [comment]
> Under across_one_type, exclude does not mean nested types are skipped. It means each nested type is checked as its own type scope (its primary declaration + its extensions in the file), separately from the parent. So Parent.Nested is still fully covered — just not merged into Parent.

Got that now. Starting with `excluded` (or whatever it's going to be called) makes a lot of sense then.

**Comment by nenadvulic** [comment]
Hi @SimplyDanny, just a follow-up whenever you get a chance, no rush at all. 
Let me know if there's anything else you'd like me to adjust. 
Thanks for your time!
